staking/slash/double-sign.go

```diff
@@ -1,52 +1,291 @@
+		Beneficiary      string   `json:"beneficiary"`
+		AddressForBLSKey string   `json:"offender"`
+	}{r.ConflictingBallots, r.Evidence, reporter, offender})
+}
+
+func (e Evidence) String() string {
+	s, _ := json.Marshal(e)
+	return string(s)
+}
+
+func (r Record) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
 
 // Verify checks that the signature is valid
 func Verify(candidate *Record) error {
+
+	fmt.Println("need to verify the slash", candidate)
+
 	return nil
 }
 
-// Apply ..
-func Apply(state *state.DB, slashes []byte) error {
+var (
+	oneHundredPercent = numeric.NewDec(1)
+	fiftyPercent      = numeric.MustNewDecFromStr("0.5")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(
+	amount *big.Int, rate numeric.Dec,
+) (*big.Int, *big.Int, *big.Int) {
+	amountPostSlash := numeric.NewDecFromBigInt(
+		amount,
+	).Mul(numeric.OneDec().Sub(rate)).TruncateInt()
+	amountOfSlash := new(big.Int).Sub(amount, amountPostSlash)
+	return amountPostSlash, amountOfSlash, new(big.Int).Div(amountOfSlash, common.Big2)
+}
+
+// postSlashBalanceCurrentDelegator applySlashRate
+// returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func postSlashBalanceCurrentDelegator(
+	state *state.DB,
+	validatorAddr common.Address,
+	delegatorAddr common.Address,
+	rate numeric.Dec,
+) (*big.Int, *big.Int, *big.Int) {
+	if wrapper := state.GetStakingInfo(validatorAddr); wrapper != nil {
+		for _, delegation := range wrapper.Delegations {
+			if delegation.DelegatorAddress == delegatorAddr {
+				return applySlashRate(delegation.Amount, rate)
+			}
+		}
+	}
+	fmt.Println("invariant violated, should not come here")
+	return nil, nil, nil
+}
+
+func delegatorSlashApply(
+	delegatedToValidator *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for _, delegationSnapshot := range delegatedToValidator.Delegations {
+		postSlashBal, slashDebt, halfOfSlashDebt := postSlashBalanceCurrentDelegator(
+			state, delegatedToValidator.Address,
+			delegationSnapshot.DelegatorAddress, rate,
+		)
+		if postSlashBal.Sign() == -1 {
+			stillOwe := new(big.Int).Sub(slashDebt, delegationSnapshot.Amount)
+			if wrapper := state.GetStakingInfo(
+				delegatedToValidator.Address,
+			); wrapper != nil {
+				for _, delegationNow := range wrapper.Delegations {
+					if delegationNow.Hash() == delegationSnapshot.Hash() {
+						for _, undelegate := range delegationNow.Undelegations {
+							// the epoch matters, only those undelegation
+							// such that epoch>= doubleSignEpoch should be slashable
+							if undelegate.Epoch.Cmp(doubleSignEpoch) == -1 {
+								continue
+							}
+							if stillOwe.Cmp(common.Big0) <= 0 {
+								// paid off the slash debt
+								break
+							}
+							amtPostSlash, amtSlash, halfOfSlashAmt := applySlashRate(
```

Comment:
Don't need to calculate slash rate again, simply take out the amount up to stillOwe, one by one over each undelegation.